rllib/offline/feature_importance.py

```diff
@@ -1,29 +1,129 @@
 from ray.rllib.utils.typing import SampleBatchType
 from ray.rllib.offline.offline_evaluator import OfflineEvaluator
 
-import numpy as np
-import copy
-
 
 @DeveloperAPI
-def perturb_fn(batch: np.ndarray, index: int):
+def _perturb_fn(batch: np.ndarray, index: int):
     # shuffle the indexth column features
     random_inds = np.random.permutation(batch.shape[0])
     batch[:, index] = batch[random_inds, index]
 
 
+@ExperimentalAPI
+def _perturb_df(batch: pd.DataFrame, index: int):
+    obs_batch = np.vstack(batch["obs"].values)
+    _perturb_fn(obs_batch, index)
+    batch["perturbed_obs"] = list(obs_batch)
+    return batch
+
+
+def _compute_actions(
+    batch: pd.DataFrame,
+    policy_state: Dict[str, Any],
+    input_key: str = "",
+    output_key: str = "",
+):
+    """A custom local function to do batch prediction of a policy.
+
+    Given the policy state the action predictions are computed as a function of
+    `input_key` and stored in the `output_key` column.
+
+    Args:
+        batch: A sub-batch from the dataset.
+        policy_state: The state of the policy to use for the prediction.
+        input_key: The key to use for the input to the policy. If not given, the
+            default is SampleBatch.OBS.
+        output_key: The key to use for the output of the policy. If not given, the
+            default is "predicted_actions".
+
+    Returns:
+        The modified batch with the predicted actions added as a column.
+    """
+    if not input_key:
+        input_key = SampleBatch.OBS
+
+    policy = Policy.from_state(policy_state)
+    sample_batch = SampleBatch(
+        {
+            SampleBatch.OBS: np.vstack(batch[input_key].values),
+        }
+    )
+    actions, _, _ = policy.compute_actions_from_input_dict(sample_batch, explore=False)
+
+    if not output_key:
+        output_key = "predicted_actions"
+    batch[output_key] = actions
+
+    return batch
+
+
+@ray.remote
+def get_feature_importance_on_index(
+    dataset: ray.data.Dataset,
+    *,
+    index: int,
+    perturb_fn: Callable[[pd.DataFrame, int], None],
+    batch_size: int,
+    policy_state: Dict[str, Any],
+):
+    """A remote function to compute the feature importance of a given index.
+
+    Args:
+        dataset: The dataset to use for the computation. The dataset should have `obs`
+            and `actions` columns. Each record should be flat d-dimensional array.
+        index: The index of the feature to compute the importance for.
+        perturb_fn: The function to use for perturbing the dataset at the given index.
+        batch_size: The batch size to use for the computation.
+        policy_state: The state of the policy to use for the computation.
+
+    Returns:
+        The modified dataset that contains a `delta` column which is the absolute
+        difference between the expected output and the output due to the perturbation.
+    """
+    perturbed_ds = dataset.map_batches(
```

Comment:
going over a same ds in parallel? this actually works? amazing.

Reply:
actually good point. I have to verify this. I forgot to update the unittest for feature_importance. added a todo for now. We should get this merged and I'll add the unittest in a separate PR after the release timeline.